src/Diagnostics/Samples/CSharp/CSharpAnalyzers/CSharpAnalyzers/StatefulAnalyzers/CodeBlockStartedAnalyzer.cs

```diff
@@ -0,0 +1,123 @@
+        internal static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.CodeBlockStartedAnalyzerDescription), Resources.ResourceManager, typeof(Resources));
+
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticIds.CodeBlockStartedAnalyzerRuleId, Title, MessageFormat, DiagnosticCategories.Stateful, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        #endregion
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCodeBlockStartAction<SyntaxKind>(UnusedParametersAnalyzer.CodeBlockStartAction);
+        }
+
+        private class UnusedParametersAnalyzer
+        {
+            #region Per-CodeBlock mutable state
+
+            private readonly HashSet<IParameterSymbol> _unusedParameters;
+            private readonly HashSet<string> _unusedParameterNames;
+
+            #endregion
+
+            #region State intialization
+
+            private UnusedParametersAnalyzer(IMethodSymbol method)
+            {
+                // Initialization: Assume all parameters are unused.
+                _unusedParameters = new HashSet<IParameterSymbol>(method.Parameters);
+                _unusedParameterNames = new HashSet<string>(method.Parameters.Select(p => p.Name));
+            }
+
+            #endregion
+
+            #region Start action
+
+            public static void CodeBlockStartAction(CodeBlockStartAnalysisContext<SyntaxKind> startCodeBlockContext)
+            {
+                // We only care about method bodies.
+                if (startCodeBlockContext.OwningSymbol.Kind != SymbolKind.Method)
+                {
+                    return;
+                }
+
+                // We only care about methods with at least one parameter.
+                var method = (IMethodSymbol)startCodeBlockContext.OwningSymbol;
+                if (method.Parameters.IsEmpty)
+                {
+                    return;
+                }
+
+                // Initialize local mutable state in the start action.
+                var analyzer = new UnusedParametersAnalyzer(method);
+
+                // Register an intermediate non-end action that accesses and modifies the state.
+                startCodeBlockContext.RegisterSyntaxNodeAction(analyzer.AnalyzeSyntaxNode, SyntaxKind.IdentifierName);
+
+                // Register an end action to report diagnostics based on the final state.
+                startCodeBlockContext.RegisterCodeBlockEndAction(analyzer.CodeBlockEndAction);
+            }
+
+            #endregion
+
+            #region Intermediate actions
+
+            private void AnalyzeSyntaxNode(SyntaxNodeAnalysisContext context)
```

Comment:
If you convert the registration to be lambda based, this would become a static method on the outer class